### PR TITLE
Some suggestions for refactoring - HAL

### PR DESCRIPTION
**Pull request by JoseClaudioSJr**
I created this draft to suggest several implementation changes in HAL

### CODE REVIEW

File: src/common.rs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+pub struct NotDefined;
+pub struct Defined;
+
+unsafe impl Send for NotDefined {}
+
+pub trait Constrain<T> {
+    fn constrain(self) -> T;
+}
+
+pub trait Split<'a> {
```
Comment (caiotavares) [inline src/common.rs]: Is it really necessary to keep this lifetime parameter here?

File: src/gpio.rs [hunk 2]
```diff
@@ -299,7 +290,7 @@ macro_rules! gpio {
             )+)+)+
         }
 
-        impl GpioExt for $DIO {
+        impl Split<'_> for $DIO {
```
Comment (caiotavares) [inline src/gpio.rs]: 🤔

File: src/clock.rs [hunk 3]
```diff
@@ -295,7 +288,7 @@ pub struct ClockConfig<MCLK, SMCLK> {
 
 macro_rules! make_clkconf {
```
Comment (caiotavares) [inline src/clock.rs:296]: Is this macro really necessary?

File: src/common.rs (hunk 1 above)
Comment (JoseClaudioSJr) [reply src/common.rs]: I will check it. We need a lifetime parameter to take a reference of the Register by the borrow function  in impl <'a>Split<'a> for TIMER32

File: src/gpio.rs (hunk 2 above)
Comment (JoseClaudioSJr) [reply src/gpio.rs]: Check the impl Split in timer32.rs.

File: src/clock.rs (hunk 3 above)
Comment (JoseClaudioSJr) [reply src/clock.rs:296]: Maybe not. The clock module has not yet been refactored.

File: src/common.rs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+//! Hal Common Resources - MSP432P401R
+
+pub struct NotDefined;
+pub struct Defined;
+
+unsafe impl Send for NotDefined {}
```
Comment (caiotavares) [inline src/common.rs]: Why is this necessary?

File: examples/timer32.rs [hunk 5]
```diff
@@ -56,119 +61,112 @@ fn main() -> ! {
     let mut p2_1 = gpio.p2_1.into_output();
     let mut p2_2 = gpio.p2_2.into_output();
 
-    let mut tim32 = p.TIMER32.constrain().set_clock(_clock);
+    let tim32 = p.TIMER32.split().set_clock(_clock);
+    let mut timer0 = tim32.channel0;
+    let mut timer1 = tim32.channel1;
+    let tim32 = tim32.tim.unwrap();
```
Comment (caiotavares) [inline examples/timer32.rs]: It's not a good practice to redefine a var like this

File: examples/timer32.rs [hunk 6]
```diff
@@ -56,119 +61,112 @@ fn main() -> ! {
...
 
     let mut count = Count(3, TimerUnit::Seconds);
 
     p1_0.try_toggle().unwrap();
 
     // TIM32 Channel 0 - One shot Mode
-    tim32.channel0().try_start_oneshot(count).unwrap();
+    timer0.borrow(&tim32).try_start_oneshot(count).unwrap();
```
Comment (caiotavares) [inline examples/timer32.rs]: Seems a bit counterintuitive to keep using tim32 for every single fn of timer0 and timer1

File: src/pmap.rs [hunk 7]
```diff
@@ -2,14 +2,13 @@
...
 pub use pac::pmap;
 use cortex_m::interrupt;
+use crate::common::Constrain;
 use crate::gpio::{Alternate, Primary};
 
 const UNLOCK_KEY: u16 = 0x2D52;
 const OFFSET0: u16 = 0x0000;
 const OFFSET1: u16 = 0x0008;
 
-static mut PORT_MAP: PmapControl = PmapControl {map: None};
```
Comment (caiotavares) [inline src/pmap.rs:11]: 👏

File: src/timer/timer32.rs [hunk 8]
```diff
@@ -1,21 +1,19 @@
...
 use core::marker::PhantomData;
-use core::mem::transmute;
 
 use crate::clock::Clocks;
 use crate::time::Hertz;
 use crate::timer::*;
 
-const MAX_PRESCALER32: u32 = 0x0100;
+unsafe impl Send for Channel0 {}
+unsafe impl Send for Channel1 {}
```
Comment (caiotavares) [inline src/timer/timer32.rs]: Why is this necessary?

File: src/common.rs (hunk 4 above)
Comment (JoseClaudioSJr) [reply src/common.rs]: Send Trait is used when transferring variables between handlers (Interrupts). In the old timer32 example, this was necessary.

"Unless these types use synchronization to manage this access, they are absolutely not thread-safe. Rust captures this through the Send and Sync traits.

A type is Send if it is safe to send it to another thread.
A type is Sync if it is safe to share between threads (T is Sync if and only if &T is Send)."

File: examples/timer32.rs (hunk 5 above)
Comment (JoseClaudioSJr) [reply examples/timer32.rs]: Ok

File: src/timer/timer32.rs (hunk 8 above)
Comment (JoseClaudioSJr) [reply src/timer/timer32.rs]: A type is Send if it is safe to send it to another thread.
